File: solutions/csharp/raindrops/Raindrops.cs
using System;
using System.Text;
using System.Collections.Generic;

public static class Raindrops
{
    private static readonly Dictionary<int, string> _drops = new Dictionary<int, string> {
        { 3, "Pling" }, { 5, "Plang" }, { 7, "Plong" } };

    public static string Convert(int number)
    {
        // determine the maximum length of all possible raindrop sounds
        int maxLength = 0;
        foreach (string sound in _drops.Values)
        {
            maxLength += sound.Length;
        }

        // build string with raindrop sounds corresponding to factors
        StringBuilder sb = new StringBuilder(maxLength);
        foreach (KeyValuePair<int, string> factor in _drops)
        {
            if (number % factor.Key == 0) { sb.Append(factor.Value); }
        }

        return (sb.Length != 0) ? sb.ToString() : number.ToString();
    }
}

File: solutions/typescript/raindrops/raindrops.ts
export function convert(num: number) {
  const isDivisbleBy = (n: number): boolean => num % n == 0
  let sound: string = ""
  if (isDivisbleBy(3)) sound += "Pling"
  if (isDivisbleBy(5)) sound += "Plang"
  if (isDivisbleBy(7)) sound += "Plong"
  return sound || num.toString()
}
